drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,572 @@ impl ServerImpl {
                 status |= Status::LOW_POWER_MODE;
             }
-            if (!presentl.get() & mask) != 0 {
+            if (!mod_status.modprsl & mask) != 0 {
                 status |= Status::PRESENT;
             }
-            if (!irql.get() & mask) != 0 {
+            if (!mod_status.intl_rxlosl & mask) != 0 {
                 status |= Status::INTERRUPT;
             }
-            if (pg.get() & mask) != 0 {
+            if (mod_status.power_good & mask) != 0 {
                 status |= Status::POWER_GOOD;
             }
-            if (pg_to.get() & mask) != 0 {
+            if (mod_status.power_good_timeout & mask) != 0 {
                 status |= Status::FAULT_POWER_TIMEOUT;
             }
-            if (pg_lost.get() & mask) != 0 {
+            if (mod_status.power_good_fault & mask) != 0 {
                 status |= Status::FAULT_POWER_LOST;
             }
             // Convert from Status -> u8 and write to the output buffer
             out[count] = status.bits();
             count += 1;
         }
-        Ok(count)
+        (count, desired_result)
     }
 
     fn select_page(
         &mut self,
         page: Page,
-        mask: FpgaPortMasks,
-    ) -> Result<(), HwError> {
+        mask: LogicalPortMask,
+    ) -> ModuleResult {
         // Common to both CMIS and SFF-8636
         const BANK_SELECT: u8 = 0x7E;
         const PAGE_SELECT: u8 = 0x7F;
 
+        let mut result = ModuleResult::default();
+
         // We can always write the lower page; upper pages require modifying
         // registers in the transceiver to select it.
         if let Some(page) = page.page() {
-            self.transceivers
-                .set_i2c_write_buffer(&[page])
-                .map_err(|_e| HwError::PageSelectWriteBufFailed)?;
-            self.transceivers
-                .setup_i2c_write(PAGE_SELECT, 1, mask)
-                .map_err(|_e| HwError::PageSelectWriteFailed)?;
-            self.wait_and_check_i2c(mask)?;
+            result =
+                result.chain(self.transceivers.set_i2c_write_buffer(&[page]));
+            result = result.chain(self.transceivers.setup_i2c_write(
+                PAGE_SELECT,
+                1,
+                mask,
+            ));
+            result = result.chain(self.wait_and_check_i2c(result.success()));
+        } else {
+            // If the request is to the lower page it is always successful
+            result =
+                ModuleResult::new(mask, LogicalPortMask(0), LogicalPortMask(0))
+                    .unwrap();
         }
 
         if let Some(bank) = page.bank() {
-            self.transceivers
-                .set_i2c_write_buffer(&[bank])
-                .map_err(|_e| HwError::BankSelectWriteBufFailed)?;
-            self.transceivers
-                .setup_i2c_write(BANK_SELECT, 1, mask)
-                .map_err(|_e| HwError::BankSelectWriteFailed)?;
-            self.wait_and_check_i2c(mask)?;
+            result =
```

Comment:
Here as well.

Reply:
Now _here_ would have been a problem since the next call to `setup_i2c_write` uses `result.success()` as its mask. Given that these are SFF devices (and won't have a `bank`), not CMIS, I don't think this is the root cause of the bug you exhibit above either.